library/core/src/marker.rs

```diff
@@ -830,3 +830,9 @@ mod copy_impls {
     #[stable(feature = "rust1", since = "1.0.0")]
     impl<T: ?Sized> Copy for &T {}
 }
+
+/// Trait to help linters look for drops that happen at unexpected times
+/// Any type whose drop has a significant side effect can impl this trait
+/// so linters can let users know if it has a potentially surprising lifetime.
+#[unstable(feature = "significant_drop", issue = "93883")]
+pub trait SignificantDrop {}
```

Comment:
hmm, i don't think we _ever_ add marker traits to the stdlib for clippy's sake, stuff is always done via adding diagnostic items, or adding a `#[clippy::foo]` attribute.

If you want this to be something others can use, I suggest doing this purely in clippy and adding a `#[clippy::has_significant_drop]` attribute (and adding it to the mutex guard if you'd like). Still okay to make the PR here.

Reply:
@Manishearth Thank you for taking the time to review this!

This got me thinking about it a little bit and maybe there's a better way to do this. I'll follow up after I've had some time to look at this more closely. Thank you!

Reply:
My reasoning for a marker trait was that we can turn it into an auto-trait so that you'll also get a warning if you return a value of `struct MyWrapper<'a>(MutexGuard<'a, u32>)`. The current impl neither has tests for it, nor will it work, nor have I communicated this plan. So yea, starting with the attribute seems like the best way forward, and then we can still decide to recurse into the type to see if any fields' types have that attribute

Reply:
@oli-obk Apologies if I'm going backwards, but would it make sense to look for any impl of Drop itself in the scrutinee that isn't borrowed or taken ownership of in an arm? I guess this will end up causing *a lot* more lint warnings to be emitted?

Reply:
Yea, if we just go with `Drop` you'll be linting `match foo.to_string().is_empty()`. Though we could see if additionally restricting to types that also are `!Freeze` (https://doc.rust-lang.org/nightly/nightly-rustc/rustc_middle/ty/struct.Ty.html#method.is_freeze) will do the trick? MutexGuard should not be `Freeze`, as it contains references to `!Freeze` types, but it would be good to check that before going down this road

Reply:
Understood! I will look at that now. Thank you!

Reply:
@Manishearth @oli-obk It didn't seem like using drop/freeze was going to work, so I used an attribute. I know this needs work in terms of improving error messages and checking other/edge cases, but I think it's working at least for the test cases in this PR. I would appreciate any further comments you might have!

Reply:
Actually, I just realized this is not passing automation. I will address this and let you know when it's passing!

Reply:
@Manishearth @oli-obk Thank you for your patience! I think this is ready for another look now.